### PR TITLE
Click tracking hook

### LINKED ISSUE #9070: Event tracking - create a hook that can send click events to ATI
**Is your feature request related to a problem? Please describe.**
We need a mechanism for sending click events for components to ATI.

**Describe the solution you'd like**
A custom hook containing the mechanism that we can add to any component to track click events.

<details open>
<summary>Acceptance Criteria</summary>

_Discuss if this can merged without attaching the hook to any components. Definition of done is unit tests that show the hook works as expected. There would be a separate issue for attaching the view/impression hook along with the click tracking hook to the most read component on the Pidgin service. Having both events working together is essential for ‘click : view’ ratio in ATI._

- A 'click' event is sent to ATI when the component is clicked
  - **Example:**
Given I view the component on the page
When I left click the component
Then I can see the event in the ATI viewer or Tag Inspector
And the values are expected<br>
Given I view the component on the page
When I right click the component
Then I don't see an event<br>
Given I view the component on the page on a Mac
When I right click the component using left click and ctrl
Then i don't see an event

We must support these click interactions:
- middle click – IE
- middle click - other than IE
- unmodified left click
- left click and shift for Mac platform
- left click and shift for Windows platform
- left click and cmd for Mac platform
- left click and ctrl for Windows platform
- left click, shift and cmd for Mac platform
- left click, shift and ctrl for Windows platform
- left click, shift, option and cmd for Mac platform
- left click, shift, alt and ctrl for Windows platform
- left click, shift and option for Mac platform
- left click, shift and alt for Windows platform
- left click, cmd and option for Mac platform
- left click, ctrl and alt for Windows platform

Example of code that can support this

```js
const leftClick = button => button === 0;
const middleClick = button => button === 1 || button === 4; // middle click for IE is 4

const isClickWithShiftOnly = event =>
  leftClick(event.button) && event.shiftKey && !(event.metaKey || event.altKey || event.ctrlKey);

const isCommonOpenClicked = event => middleClick(event.button) || isClickWithShiftOnly(event);

const isSupportedClickModifier = (event, functionKey) =>
  (functionKey && !event.shiftKey && !event.altKey) || // functionKey
  (event.shiftKey && functionKey && !event.altKey) || // shift + functionKey
  (event.shiftKey && functionKey && event.altKey) || // shift + functionKey + option/alt
  (event.shiftKey && event.altKey && !functionKey) || // shift + option/alt
  (event.altKey && functionKey && !event.shiftKey); // option/alt + cmd

const isMacOsOpenClicked = event => leftClick(event.button) && isSupportedClickModifier(event, event.metaKey); // cmd

const isWindowsOpenClicked = event => leftClick(event.button) && isSupportedClickModifier(event, event.ctrlKey); // ctrl

const isMacOs = () => window && window.navigator && window.navigator.platform.startsWith('Mac');

export const isSafari = () => window.navigator.vendor && window.navigator.vendor.startsWith('Apple');

export const isOpenClicked = event =>
  (isMacOs() ? isMacOsOpenClicked(event) : isWindowsOpenClicked(event)) || isCommonOpenClicked(event);

export const isNotModifiedLeftClick = event =>
  leftClick(event.button) && !(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
```

**Example ATI event**
```
From page level call (from pidgin)
s represents WS_NEWS_LANGUAGES and s2 represents the site

https://a1.api.bbc.co.uk/hit.xiti?
&s=598342
&idclient=3344b414-096a-4e3a-9c44-2ebd2066a0ad
&r=1920x1080x24x24
&re=844x969
&hl=16x35x53
&lng=en-GB
&p=pidgin.page
&s2=70

Additional field for click event (ati)
&ati=PUB-[WSTestEvent]-[TestClick]-[variant]-[format]-[general]-[detailed]-[adverts]-[url]
&type=AT 
```
</details>

<details open>
<summary>Additional Info</summary>
There is a prototype custom hook that logs click events to the browser https://github.com/bbc/simorgh/pull/9064

The event logs to the browser console will need to be replaced with an actual call to ATI. There was a spike around how to do this in Simorgh. The outcome is a prototype PR that can be referred to for sending the events to ATI https://github.com/bbc/simorgh/pull/9063

🐛 There is an issue with ATI for Safari where navigating back after clicking on a tracked link triggers the link again - story BBCHOME-398. We must consider a workaround for this. WebCore have fixed this in their implementation - refer to their workaround for this.

We need to prevent navigations on click before logging the event - refer to Reverb/ATI code regarding preventing navigation before event despatch.
</details>

**Describe alternatives you've considered**

</details>

<details open>
<summary>Questions</summary>

- Do we need to remove the previous click tracking mechanism in Simorgh?
- Do we log only one click when a user double clicks?
- Should we be concerned about clicks vs taps?
- a11y - what about keypress enter event - ask Catharine
</details>

**Testing notes**
[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [ ] This feature is expected to need manual testing.

**Additional context**
Story - NEWSWORLDSERVICE-1256

### PR DESCRIPTION
**Pull request by HarryVerhoef**
Resolves https://github.com/bbc/simorgh/issues/9070

**Overall change:**
Added a custom hook that can be used to send view and click events to ATI on click.

**Code changes:**

- Created a custom hook, `useClickTracker`, that returns a ref which can be given to any component.
- Added some assertions that test the hook functions as expected.
- Added variant as another parameter to `buildATIClickParams`.

---

- [X] I have assigned myself to this PR and the corresponding issues
- [X] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [X] I have assigned this PR to the Simorgh project

**Testing:**

- [X] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/containers/ATIAnalytics/atiUrl/index.js [hunk 1]
```diff
@@ -233,6 +234,12 @@ export const buildATIEventTrackUrl = ({
 
   const eventPublisher = type === 'view' ? 'ati' : 'atc';
   const eventTrackingBeaconValues = [
+    {
+      key: 'idclient',
+      description: 'at user id',
+      value: getAtUserId(),
+      wrap: false,
+    },
```
Comment (andrewscfc) [inline src/app/containers/ATIAnalytics/atiUrl/index.js]: I'm sure you are aware but the same changes have been made in @jroebu14 's PR so there may be a mild conflict whichever is merged first

File: src/app/hooks/useClickTracker/helpers.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+const isLeftClick = button => button === 0;
```
Comment (andrewscfc) [inline src/app/hooks/useClickTracker/helpers.js]: Has this been lifted directly from the Web Core implementation? I wonder if we should offer to maintain this as a utility library in psammead for them to use rather than duplicating the code? Just a thought, I wouldn't block this PR.

File: src/app/hooks/useClickTracker/helpers.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  isCommonOpenClicked(event);
+
+export const isNotModifiedLeftClick = event =>
+  isLeftClick(event.button) &&
+  !(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
+
+export const isValidClick = event =>
+  isMiddleClick(event.button) ||
+  isNotModifiedLeftClick(event) ||
+  isOpenClicked(event);
```
Comment (andrewscfc) [inline src/app/hooks/useClickTracker/helpers.js]: We may want to go through our WS browser support list and see if we have full support here for all our browsers e.g. Opera? Maybe a better one for testing @paruchurisilpa @JackJaeger and a fix forward later if anything is found

File: src/app/hooks/useClickTracker/index.jsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  const handleClick = useCallback(
+    event => {
+      if (!hasBeenClicked && isValidClick(event)) {
+        setHasBeenClicked(true);
+        const componentInfo = getComponentInfo({
+          result: event.target.href || window.location.href,
+          componentName,
+          componentData: {
+            actionLabel: 'click',
+            child: event.target.tagName,
```
Comment (andrewscfc) [inline src/app/hooks/useClickTracker/index.jsx]: has this been verified as correct with @JonBeeb , @jroebu14  wasn't sure what to provide here on his viewing tracking PR?

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  return (
+    <div>
+      <TestComponent ref={clickRef} />
+    </div>
+  );
+};
+
+const TestComponent = React.forwardRef((_, ref) => {
+  return (
+    <div id="test-component" ref={ref}>
```
Comment (andrewscfc) [inline src/app/hooks/useClickTracker/index.test.jsx]: Interesting will this make the click tracking component track clicks on:
- the div?
- the link?
- the button?
- all of them?

The component probably needs the flexibility to be used on either a containing element such as a `div` or a lower level element such as an `<a>` or `<button>`.

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+      wrapper,
+    });
+
+    const requestContext = renderHook(() => useContext(RequestContext), {
+      wrapper,
+    });
+
+    const serviceContext = renderHook(() => useContext(ServiceContext), {
+      wrapper,
+    });
```
Comment (andrewscfc) [inline src/app/hooks/useClickTracker/index.test.jsx]: Would you ever use the hook with just `RequestContext` or `ServiceContext`? Could just have one call here providing both contexts? Or am I misunderstanding the `renderHook` testing API?

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 5 above)
Comment (andrewscfc) [reply src/app/hooks/useClickTracker/index.test.jsx]: I see below you test for this to some extent, would we want the flexibility to track clicks on the containing `div` and seperately track clicks on a contained element like the `<a>` or the `<button>`? So seperate click refs for the `div` and the `<button>`/`<a>`? Perhaps a questions for @JonBeeb and maybe worth a test in here if it's necessary?

**Review by andrewscfc — COMMENTED** [review]
Looks great, just some stuff to think about 👍

**Comment by andrewscfc** [comment]
Oh forgot some stuff from our discussion with Web Core last week:
- Is there a need for logic in here to prevent navigation before the click is dispatched? I can't see anything in the PR and I think it is probably required.
  - If we do this we need to be mindful of issues experienced with delaying the navigation in Safari that Web Core experienced
  - We also need to check in our WS browsers that delaying the navigation doesn't cause issues

I also assume the combination of view and click to maintain a click:view ratio is coming up in a later PR?

File: src/app/containers/ATIAnalytics/atiUrl/index.js (hunk 1 above)
Comment (HarryVerhoef) [reply src/app/containers/ATIAnalytics/atiUrl/index.js]: Thanks for the review @andrewscfc!

Yeah I'll bear that in mind, just had to make the same changes in order to get it to work 👍

File: src/app/hooks/useClickTracker/helpers.js (hunk 2 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/helpers.js]: Yeah it has, I added a few methods and changed some naming though. I'll ask Inti 👍

File: src/app/hooks/useClickTracker/index.jsx (hunk 4 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.jsx]: These haven't been verified as such yet, they're mostly placeholders. That being said, I think they're appropriate values, if a user clicks on a brand by clicking the svg, it'll send `brand-click` to ati, also specifying that the svg was the child element that was actually clicked on.

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 5 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx]: Once a click event listener is attached to an element, any element which is a child of that parent element can be clicked and the listener will fire. In the ati request it specifies the actual element that has been clicked on, whether that be a button or an anchor 👍

Comment (andrewscfc) [reply src/app/hooks/useClickTracker/index.test.jsx]: Ah okay, I don't think you'd need this but a slightly weird edge case I was consider was:
- You want track that a containing component has been clicked, e.g. a collapsible panel that reveals some more content (https://www.bbc.co.uk/gel/guidelines/accordion)
- You then want track if a button inside that revealed content has also been clicked independently

I don't think we have anything like this but just came to mind. With this current solution I suspect you could only track the collapsible panel had been clicked.

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 6 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx]: I haven't much experience with the `renderHook` API myself. I just tried this:
```js
const contexts = renderHook(() => {
    return {
      request: useContext(RequestContext),
      service: useContext(ServiceContext),
    };
});
```
It doesn't seem to get a value for either though, how were you thinking of combining the two?

**Comment by HarryVerhoef** [comment]
> * Is there a need for logic in here to prevent navigation before the click is dispatched? I can't see anything in the PR and I think it is probably required.

> I also assume the combination of view and click to maintain a click:view ratio is coming up in a later PR?

Yeah I still need to look into this, just thought I'd get this one up for review as a sort of foundational click tracker.
@jroebu14 and I are going to discuss potentially combining of the two at some point 👍

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 5 above)
Comment (jroebu14) [reply src/app/hooks/useClickTracker/index.test.jsx]: In this case I think we need to consider calling stopPropagation on the event.

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 7]
```diff
@@ -0,0 +1,174 @@
+/* eslint-disable react/prop-types */
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx:2]: I wrote some tests around error handling to ensure the view tracker hook doesn't break the app when unexpected data is passed to it https://github.com/bbc/simorgh/pull/9071/files#diff-81136bbf60a2de9e097e285911100a0850d8b5dc7985bee4376851cf43e47da8R187

You could literally copy/paste these and change the tests to use your hook and then make those tests pass by changing your hook code.

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        <TestComponent ref={result.current} />
+      </WithContexts>,
+    );
+
+    expect(result.current.current).toBe(
+      container.querySelector('#test-component'),
+    );
+  });
+
+  it('should call buildATIClickParams with page data, requestContext, serviceContext', () => {
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx]: I don't think I see much value in this unit test. We could easily refactor your custom hook so it replaces requestContext, serviceContext and buildATIClickParams with some other implementation to successfully make calls to ATI on click events. This would mean the hook works fine but this test would say it's broken so you'd have to then spend time refactoring the tests to pass again even though you didn't actually break the hook functionality. It doesn't really matter that these functions were called. A test like this is kind of like pouring concrete around the code and making it difficult to refactor. I'm only saying this because I think we probably will want to refactor the ATI functions in the future. This article around not testing implementation details is worth a read https://kentcdodds.com/blog/testing-implementation-details also I'm more than happy to explain on a call!

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    });
+
+    expect(params.buildATIClickParams).toHaveBeenCalledWith(
+      defaultProps.pageData,
+      requestContext.result.current,
+      serviceContext.result.current,
+    );
+  });
+
+  it('should call sendEventBeacon on click', async () => {
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx]: Same as the above comment I made - not sure we need this. I think the test `should send tracking request on click of child element (button)` is the kind of unit test that adds the most value and confidence and allows refactors of code in the future.

**Review by jroebu14 — COMMENTED** [review]
This is looking great. I wonder if we should have a discussion about stopPropagation on the event on Tuesday though as I think it relates to a point @andrewscfc made. WebCore are using stopPropagation in some situations in their code.

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 8 above)
Comment (jroebu14) [reply src/app/hooks/useClickTracker/index.test.jsx]: I just thought of this https://github.com/bbc/simorgh/pull/9071#discussion_r621287011

If I renamed buildATIClickParams function and filename and then updated the naming in view tracker and click tracker, things would still work in both hooks but this test would be broken meaning I have the extra chore of updating the test.

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 7 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx:2]: Awesome, thanks, I'll do that now 👍

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 8 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx]: Ahhh that makes a lot of sense. I'll remove the test and give that article a read, thanks!

**Comment by HarryVerhoef** [comment]
> This is looking great. I wonder if we should have a discussion about stopPropagation on the event on Tuesday though as I think it relates to a point @andrewscfc made. WebCore are using stopPropagation in some situations in their code.

Thanks @jroebu14!
I wrote a test which I think mimics the test case @andrewscfc was thinking of, and adding `stopPropagation` makes the test pass 👍

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+    const spyFetch = jest.spyOn(global, 'fetch');
+    const { container } = render(
+      <WithContexts>
+        <TestComponentContainer hookProps={defaultProps} />
+      </WithContexts>,
+    );
+
+    expect(spyFetch).toHaveBeenCalledTimes(0);
+
+    act(() => container.querySelector('#test-component').click());
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx]: I think it might be better to use the query and event utilities provided by react-tesing-library e.g.

```jsx
const { getByText } = render(
  <WithContexts>
    <TestComponentContainer hookProps={defaultProps} />
  </WithContexts>,
);

const linkEl = getByText('Test Link')
const buttonEl = getByText('Test  Button')
```

```js
import userEvent from '@testing-library/user-event'

act(() => userEvent.click(linkEl));

act(() => userEvent.click(buttonEl));
```
More on react-testing-library queries and events:
- https://testing-library.com/docs/queries/about/
- https://testing-library.com/docs/ecosystem-user-event/#clickelement-eventinit-options

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+import zhongwenData from './fixtureData/chinese-news-49631219-trad.json';
+import { RequestContextProvider } from '#contexts/RequestContext';
+import { ServiceContextProvider } from '#contexts/ServiceContext';
+import { STORY_PAGE } from '#app/routes/utils/pageTypes';
+
+beforeEach(() => {
+  jest.clearAllMocks();
+});
+
+jest.useFakeTimers();
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx]: I don't think you need this unless you're using setTimeout

Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx]: Ah yeah forgot to remove this, nice spot 👍

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 10 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx]: Makes sense, will do this now

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+      wrapper,
+    });
+
+    const { getByText } = render(
+      <WithContexts>
+        <TestComponent ref={result.current} />
+      </WithContexts>,
+    );
+
+    expect(result.current.current).toBe(getByText('Containing Div'));
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx]: sorry to be a pain but this selector might be better off using https://testing-library.com/docs/queries/bytestid rather than having to add text to the DOM to query

File: src/app/hooks/useClickTracker/helpers.js (hunk 2 above)
Comment (intih) [reply src/app/hooks/useClickTracker/helpers.js]: That sounds interesting -- would be happy with a utility library that gives us less to support 😁 It'd be interesting to see if your browser support list means you'll need to tweak this further (and if so, how much) compared to WebCore's browser support list.

Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/helpers.js]: @intih Awesome, I can make this into a utility library in psammead at some point, perhaps at the same time I investigate if any changes need to be made to cater for all the other browsers 👍

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 12 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx]: Yeah good idea, done 👍

File: src/app/hooks/useClickTracker/index.jsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  const clickRef = useRef(null);
+
+  try {
+    eventTrackingProps = buildATIClickParams(
+      pageData,
+      requestContext,
+      serviceContext,
+    );
+  } catch (e) {
+    eventTrackingProps = null;
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.jsx]: As pointed out by @andrewscfc in my PR we should probably log something to the browser console until we have client side logging in place.

This is what I did https://github.com/bbc/simorgh/pull/9071/files#diff-7847911c4a756dfc4dce2205ea165af35dbdbb6c296083a0a02862757737cb92R41-R43

and then I ensured we have logging in the unit tests https://github.com/bbc/simorgh/pull/9071/files#diff-81136bbf60a2de9e097e285911100a0850d8b5dc7985bee4376851cf43e47da8R211-R215

**Review by jroebu14 — APPROVED** [review]
Great work @HarryVerhoef 

Could do with a quick [readme like I did for the view tracker](https://github.com/bbc/simorgh/blob/view-tracking/src/app/hooks/useViewTracker/README.md) but other than this looks good.

File: src/app/hooks/useClickTracker/index.jsx [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        if (eventTrackingProps && componentInfo) {
+          sendEventBeacon({
+            type: 'click',
+            componentName,
+            service,
+            componentInfo,
+            ...eventTrackingProps,
+          })
+            .then(() => {
+              if (href || event.target.href) {
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.jsx]: this could be a constant e.g. `const nextPageUrl = href || event.target.href;` or something similar

File: src/app/hooks/useClickTracker/index.jsx [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+            componentName,
+            format,
+            pageIdentifier,
+            platform,
+            service,
+            statsDestination,
+            url: window.location.href,
+          })
+            .then(() => {
+              changeUserLocationIfDefined(nextPageUrl);
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.jsx]: I wonder if the call to this function could exist in a `finally` block https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally

Then you wouldn't have to create this function at all e.g.

```js
          sendEventBeacon({
            type: EVENT_TYPE,
            campaignName,
            componentName,
            format,
            pageIdentifier,
            platform,
            service,
            statsDestination,
            url: window.location.href,
          })
            .catch(error => {
              console.error(
                `Error sending ATI click tracking request: ${error.message}`,
              );
            })
            .finally(() => {
              if (url) {
                  window.location.assign(url);
              }
            });
```

Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.jsx]: Awesome! Cheers @jroebu14

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+
+  return {
+    origin,
+    pathname,
+    searchParams: Object.fromEntries(searchParams),
+  };
+};
+
+delete window.location;
+window.location = { href: 'http://bbc.com/pidgin/tori-51745682' };
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx]:
```suggestion
window.location = {
  href: 'http://bbc.com/pidgin/tori-51745682',
  assign: jest.fn(),
};
```

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+    });
+
+    window.location.assign = jest.fn(href => {
+      try {
+        expect(href).toBe(url);
+        done();
+      } catch (error) {
+        done(error);
+      }
+    });
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx]: I'm not sure this is working as we expect it to. You can mock `assign` like this:

https://github.com/bbc/simorgh/pull/9072/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.jsx&file-filters%5B%5D=.lock#r629124795

then you can check it was called like this:

```js
    await waitFor(() => {
      expect(window.location.assign).toHaveBeenCalledWith(
        'https://bbc.com/pidgin',
      );
    });
```

note that it needs to be in a `waitFor` because `window.location.assign` is within a promise

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 16 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx]: Yeah I tried this approach but mocking the assign method is proving to be really tricky, I know it's getting called from manual testing but can't see it from the automated tests. Only way I can get it to pass is to do the assertion within a mock implementation of the assign method. Would you be able to hop on a call at some point today so I can better explain this?

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+      },
+    });
+
+    jest.restoreAllMocks();
+  });
+
+  it('should allow the user to navigate after clicking on a tracked link even if the tracking request fails', done => {
+    const url = 'https://bbc.com/pidgin';
+
+    const spyFetch = jest.spyOn(global, 'fetch').mockImplementation(() => {
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx]: global fetch is already mocked in a `jest-setup` file that runs before all tests https://github.com/bbc/simorgh/blob/latest/src/testHelpers/jest-setup.js#L22

so you can just mock it again 

```js
    global.fetch = jest.fn(() => {
      throw new Error('Failed to fetch');
    });
```

`jest.fn(() => {` is a another way to write `mockImplementation` here

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 16 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx]: Actually, never mind, your comment below has fixed it, cheers!

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 17 above)
Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/index.test.jsx]: Legend, that's worked a treat, thanks!

**Review by jroebu14 — COMMENTED** [review]
Sorry for the never ending suggestions. Just a couple more things I've noticed

File: src/app/hooks/useClickTracker/index.test.jsx (hunk 16 above)
Comment (jroebu14) [reply src/app/hooks/useClickTracker/index.test.jsx]: No worries! I'm happy to explain on a call if you like. Also, a bit of an oversight in the view hook too, but we should restore the original functionality of window.location when we've done testing - just to be sure there are no side-effects when people add more tests. Something like this would work https://remarkablemark.org/blog/2018/11/17/mock-window-location/#solution

```js
  const { location } = window;

  beforeAll(() => {
    delete window.location;
    window.location = {
      href: 'http://bbc.com/pidgin/tori-51745682',
      assign: jest.fn(),
    };
  });

  afterAll(() => {
    window.location = location;
  });
```

File: src/app/hooks/useClickTracker/index.test.jsx [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+  });
+
+  it('should allow the user to navigate after clicking on a tracked link even if the tracking request fails', async () => {
+    const url = 'https://bbc.com/pidgin';
+
+    global.fetch = jest.fn(() => {
+      throw new Error('Failed to fetch');
+    });
+
+    window.location.assign = jest.fn();
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/index.test.jsx]: I don't think we need this line

File: src/app/hooks/useClickTracker/README.md [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+};
+```
+
+### Props
+
+| Argument      | Type   | Required | Example                                                                                                                              |
+| ------------- | ------ | -------- | ------------------------------------------------------------------------------------------------------------------------------------ |
+| pageData      | object | yes      | The page data used to hydrate the page.                                                                                              |
+| componentName | string | yes      | The name of the component or an url encoded title of a promo e.g. `most_read` or `This%20is%20a%20promo%20title`.                    |
+| campaignName  | string | yes      | The name of the campaign e.g. `cps_wsoj`                                                                                             |
```
Comment (jroebu14) [inline src/app/hooks/useClickTracker/README.md]:
```suggestion
| campaignName  | string | yes      | The name of the campaign e.g. `cps_wsoj` typically defined by Business Analyists or Product Owners in the analytics interface |
```

**Review by jroebu14 — APPROVED** [review]
couple of small suggestions but this is looking really nice!

File: src/app/hooks/useClickTracker/README.md [hunk 21]
```diff
@@ -0,0 +1,57 @@
+# useClickTracker hook
+
+The `useClickTracker` hook handles:
+
+- Tracking when an element has been clicked
+- Sending the event to ATI
+
+A click event is sent to ATI when a user performs a valid click (as per [clickTypes.js](./clickTypes.js)) on a tracked component. Specificially the following are valid clicks:
```
Comment (ryanmccombe) [inline src/app/hooks/useClickTracker/README.md]:
```suggestion
A click event is sent to ATI when a user performs a valid click (as per [clickTypes.js](./clickTypes.js)) on a tracked component. Specifically the following are valid clicks:
```

File: src/app/hooks/useClickTracker/README.md [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+};
+```
+
+### Props
+
+| Argument      | Type   | Required | Example                                                                                                                              |
+| ------------- | ------ | -------- | ------------------------------------------------------------------------------------------------------------------------------------ |
+| pageData      | object | yes      | The page data used to hydrate the page.                                                                                              |
+| componentName | string | yes      | The name of the component or an url encoded title of a promo e.g. `most_read` or `This%20is%20a%20promo%20title`.                    |
+| campaignName  | string | yes      | The name of the campaign e.g. `cps_wsoj` typically defined by Business Analyists or Product Owners in the analytics interface |
```
Comment (ryanmccombe) [inline src/app/hooks/useClickTracker/README.md]:
```suggestion
| campaignName  | string | yes      | The name of the campaign e.g. `cps_wsoj` typically defined by Business Analysts or Product Owners in the analytics interface |
```

File: src/app/hooks/useClickTracker/clickTypes.js [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+  isLeftClick(event.button) && isSupportedClickModifier(event, event.metaKey); // cmd
+
+const isWindowsOpenClicked = event =>
+  isLeftClick(event.button) && isSupportedClickModifier(event, event.ctrlKey); // ctrl
+
+const isMacOs = () =>
+  window && window.navigator && window.navigator.platform.startsWith('Mac');
+
+export const isSafari = () =>
+  window.navigator.vendor && window.navigator.vendor.startsWith('Apple');
```
Comment (ryanmccombe) [inline src/app/hooks/useClickTracker/clickTypes.js]: Are we sure this function is never called on the server, where `window` will be undefined?

Comment (HarryVerhoef) [reply src/app/hooks/useClickTracker/clickTypes.js]: @ryanmccombe This will only ever be run as part of a click event handler so I'm pretty sure it won't be called on the server, but I can add `window` as a condition just in case 👍

**Review by ryanmccombe — APPROVED** [review]
Nice 👍

File: src/app/hooks/useClickTracker/clickTypes.js (hunk 23 above)
Comment (ryanmccombe) [reply src/app/hooks/useClickTracker/clickTypes.js]: Cool :) Looks good!

Ramda is pretty good for safely handling these deep tests too: 

```javascript
pathSatisfies(startsWith('Mac'), ['navigator', 'platform'], window)
```

https://ramdajs.com/docs/#pathSatisfies
https://ramdajs.com/docs/#startsWith
